### PR TITLE
Feat/tensorboard

### PR DESCRIPTION
**Pull request by ahmedheakl**
Feature and Fix - Tensorboard and Optimizing the timing for difficulty

## 📜 Summary
Since we need to keep track of the stats for the training, tensorboard will provide a way to monitor the training in real-time. Moreover, some training sessions for the robot used the difficulty checker before checking for obstacles overlapping, hence leading to marginally increasing the training time. In this PR, we integrate tensorboard to view training stats for both the teacher and the robot, and optimize the difficulty checker module for faster training. 

###### Ticket(s): [Tensorboard Integration](https://app.clickup.com/t/8669g6cdk)

- 🔄 **Main changes in this PR**
    - Added logging using `torch.tensorboard.SummaryWriter` to both teacher and robot
    - Modified the overlapping checking to run before the `convex_hull_difficulty` for optimizing the training.
    - Corrected difficulty value assignment

- ⚠️ **Breaking changes**
  - None

## 🚀 How to test this Pull Request

1. `pytest test`

## 📝 Notes for future improvements

- Optimize the implementation for the `teacher_checker` module.

### CODE REVIEW

File: src/highrl/envs/robot_env.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
 from highrl.utils.calculations import point_to_point_distance
 from highrl.agents.robot import Robot
 from highrl.obstacle.obstacles import Obstacles
+from torch.utils.tensorboard import SummaryWriter
```
Comment (ahmedheakl) [inline src/highrl/envs/robot_env.py:31]: Make relative import the last.
```python
from torch.utils.tensorboard import SummaryWriter

from highrl.utils.calculations import point_to_point_distance
from highrl.agents.robot import Robot
from highrl.obstacle.obstacles import Obstacles
```

File: src/highrl/envs/robot_env.py [hunk 2]
```diff
@@ -101,6 +102,8 @@ def __init__(
                     "wall_time",
                 ]
             )
+        self.robot_episode_reward_writer = SummaryWriter(log_dir="runs")
```
Comment (ahmedheakl) [inline src/highrl/envs/robot_env.py:105]: Create global variable for folder path instead for hard coding the value, for future easy manipulation.

File: src/highrl/envs/robot_env.py [hunk 3]
```diff
@@ -168,6 +171,16 @@ def step(self, action: List) -> Tuple:
         )
 
         self.episode_reward += self.reward
+        self.robot_reward_writer.add_scalar(
```
Comment (ahmedheakl) [inline src/highrl/envs/robot_env.py:174]: Same here with the hardcoding.

File: src/highrl/envs/teacher_env.py [hunk 4]
```diff
@@ -27,6 +27,7 @@
 from prettytable import PrettyTable
 import argparse
 from os import path
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:29]: `prettytable` library import should be place after builtin import `os` and `argparse`.

File: src/highrl/envs/teacher_env.py [hunk 5]
```diff
@@ -98,6 +99,13 @@ def __init__(
                     "robot_num_successes",  # robot num_successes in this teacher session
                 ]
             )
+        self.teacher_reward_writer = SummaryWriter(log_dir="runs")
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:102]: Same with hardcoding

File: src/highrl/envs/teacher_env.py [hunk 6]
```diff
@@ -200,6 +208,21 @@ def _get_robot_metrics(self) -> None:
                     f"{self.robot_success_rate:0.2f}",
                 ]
             )
+            self.robot_avg_reward_writer.add_scalar(
+                "robot_avg_reward",
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:212]: Same with hardcoding

File: src/highrl/envs/teacher_env.py [hunk 7]
```diff
@@ -251,6 +267,19 @@ def step(self, action: List) -> Tuple:
                 )
             )
         is_goal_or_robot_overlap_obstacles = is_goal_or_robot_overlap_obstacles > 0
+        if is_goal_overlap_robot or is_goal_or_robot_overlap_obstacles:
+            self.difficulty_area = self.INF_DIFFICULTY
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:271]: Modify this to be: 
```python
if is_goal_overlap_robot: 
   self.difficulty_area = 0
   self.difficulty_obs = 0

elif is_goal_or_robot_overlap_obstacles:
  self.difficulty_area = INF_DIFFICULTY
  self.difficulty_obs = 1
```

File: src/highrl/envs/teacher_env.py [hunk 8]
```diff
@@ -279,6 +308,21 @@ def step(self, action: List) -> Tuple:
                     self.robot_env.num_successes,
                 ]
             self.penality_time_step = self.time_steps
+            self.teacher_reward_writer.add_scalar(
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:311]: Same with hardcoding

File: src/highrl/envs/teacher_env.py [hunk 9]
```diff
@@ -305,7 +349,7 @@ def step(self, action: List) -> Tuple:
             self.robot_id += 1
             # fmt: off
             print("initiating model ...")
-            model = PPO("MultiInputPolicy", self.robot_env, policy_kwargs=policy_kwargs, verbose=2, device=device)
+            model = PPO("MultiInputPolicy", self.robot_env, policy_kwargs=policy_kwargs, verbose=2, device=device,)
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:352]: Run `black` styling here to reduce the line length as follows
```python
model = PPO("MultiInputPolicy",
self.robot_env,
policy_kwargs=policy_kwargs,
verbose=2,
device=device,
)
```

File: src/highrl/envs/teacher_env.py [hunk 10]
```diff
@@ -361,6 +405,22 @@ def step(self, action: List) -> Tuple:
                 self.robot_env.num_successes
             ]
 
+            self.teacher_reward_writer.add_scalar(
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:408]: Same with hardcoding

**Review by ahmedheakl — COMMENTED** [review]
Looks great, modify requested changes and merge.

File: src/highrl/envs/teacher_env.py [hunk 11]
```diff
@@ -98,6 +99,13 @@ def __init__(
                     "robot_num_successes",  # robot num_successes in this teacher session
                 ]
             )
+        self.teacher_reward_writer = SummaryWriter(log_dir="runs")
+        self.robot_num_successes_writer = SummaryWriter(log_dir="runs")
+        self.robot_avg_reward_writer = SummaryWriter(log_dir="runs")
```
Comment (abdualamer) [inline src/highrl/envs/teacher_env.py:104]: I think this is the same like the one in the robot_env file

**Review by abdualamer — APPROVED** [review]
LGTM!!

File: src/highrl/envs/robot_env.py (hunk 2 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/robot_env.py:105]: Done

File: src/highrl/envs/robot_env.py (hunk 3 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/robot_env.py:174]: I do not understand which part exactly is in need of change?

Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/robot_env.py:174]: Done

File: src/highrl/envs/teacher_env.py (hunk 4 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/teacher_env.py:29]: Done

File: src/highrl/envs/teacher_env.py (hunk 6 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/teacher_env.py:212]: Done

File: src/highrl/envs/teacher_env.py (hunk 7 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/teacher_env.py:271]: instead of which lines?

File: src/highrl/envs/teacher_env.py (hunk 10 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/teacher_env.py:408]: Done

File: src/highrl/envs/teacher_env.py (hunk 9 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/teacher_env.py:352]: Done

File: src/highrl/envs/teacher_env.py (hunk 8 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/teacher_env.py:311]: Done

File: src/highrl/envs/teacher_env.py (hunk 11 above)
Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/teacher_env.py:104]: I do not understand your comment

File: src/highrl/envs/teacher_env.py (hunk 7 above)
Comment (ahmedheakl) [reply src/highrl/envs/teacher_env.py:271]:
```python
if is_goal_overlap_robot or is_goal_or_robot_overlap_obstacles:
    self.difficulty_area = self.INF_DIFFICULTY
    self.difficuly_obs = ....
```

Comment (AbdelrahmanAbdelgwad) [reply src/highrl/envs/teacher_env.py:271]: Done, but please revise carefully.

File: src/highrl/envs/teacher_env.py [hunk 12]
```diff
@@ -84,6 +85,14 @@ def __init__(
         self.robot_id = 0
         self.penality_time_step = 0
 
+        self.tensorboard_dir = "runs"
```
Comment (ahmedheakl) [inline src/highrl/envs/teacher_env.py:88]: To separate the tensorboard logs for the `teacher` and `robot` modules, rewrite this as
```python
self.tensorboard_dir = "runs/teacher"
```

File: src/highrl/envs/robot_env.py [hunk 13]
```diff
@@ -73,7 +74,9 @@ def __init__(
         self.robot_goal_py: int = 0
 
         self.is_initial_state = True
-
+        self.tensorboard_dir = "runs"
```
Comment (ahmedheakl) [inline src/highrl/envs/robot_env.py:77]: To separate the tensorboard logs for the teacher and robot modules, rewrite this as
```python
self.tensorboard_dir = "runs/robot"
```

**Review by ahmedheakl — COMMENTED** [review]
Great! Just alter the values of the tensorboard directory for the teacher and the robot, and you're good to go.

**Pull request merged**